lib/phoenix/channel/driver.ex

```diff
@@ -0,0 +1,252 @@
+# TODO(sj): move the file to the proper place
+defmodule Phoenix.Socket.Driver do
+  @moduledoc """
+  Implementation of the Phoenix Channels protocol.
+
+  The Phoenix Channels protocol describes how to hold multiple separate conversations
+  (channels) over a single persistent connection (socket). The connection is
+  powered by transports, such as websocket or long polling, while this module
+  provides the stock implementation of the protocol, with following properties:
+
+    * Each channel runs in a separate process.
+
+    * Channel processes are direct children of the socket process.
+
+    * Termination of a channel process doesn't affect other channel processes or
+      the socket process.
+
+    * If the socket process terminates, all channel processes will be stopped as well,
+      regardless of the exit reason.
+
+  This module was not meant to be used directly. Instead you can pass it as an option
+  to the transport (TODO(sj): explain how, once we settle on the approach). This
+  happens by default when your module is powered by `Phoenix.Socket`.
+  """
+  @behaviour Phoenix.Transports.Driver
+
+  require Logger
+  alias Phoenix.Socket
+  alias Phoenix.Socket.Broadcast
+  alias Phoenix.Socket.Message
+  alias Phoenix.Socket.Reply
+
+  @client_vsn_requirements "~> 1.0"
+  @protocol_version "1.0.0"
+
+  @doc """
+  Returns the Channel Transport protocol version.
+  """
+  def protocol_version, do: @protocol_version
+
+  @doc false
+  def transports(endpoint, driver_opts) do
+    alias Phoenix.Transports
+
+    socket_handler = Keyword.fetch!(driver_opts, :socket_handler)
+    path = Keyword.fetch!(driver_opts, :path)
+
+    for {transport_name, {transport, transport_opts}} <- socket_handler.__transports__ do
+      {
+        Path.join(path, Atom.to_string(transport_name)),
+        %Transports.Driver.Config{
```

Comment:
OK, so here's the breakdown.

Transport layer (including the cowboy handler) requires:

- full path
- transport module
- transport options


The socket driver `transports` function needs:

- endpoint
- socket module
- base path

The socket driver `init` function needs:

- socket module
- serializer module
- transport name
- transport module

I'd like to see the mandatory stuff somehow mandated, rather than letting them be defined by conventions and relying on `fetch!`. 

We could thus, separate driver options, driver init options, and transport options. In other words, the `transport` callback would return `[{path, transport_module, transport_opts, driver_init_opts}]`, and these would be then passed through the transport layer. The driver `init` callback would receive `endpoint`, `driver_init_opts`, and `params`.

That way we're more explicit about what's mandated while there's no coupling of unrelated data in the same struct. What do you think?